### PR TITLE
bls: Avoid needless hashing

### PR DESCRIPTION
**Pull request by UdjinM6**
I might be missing smth but I wasn't able to find any reason to actually update the cached hash in `CBLSWrapper` that often. IMO doing this in `GetHash()` only and only when it's actually needed should be enough.

#4012:
```
# Benchmark, evals, iterations, total, min, max, median
BLSDKG_BuildQuorumVerificationVectors_parallel_10, 5, 3000, 3.59811, 0.000229663, 0.000268742, 0.000234689
BLSDKG_BuildQuorumVerificationVectors_parallel_100, 5, 50, 4.30701, 0.0163193, 0.0198587, 0.0166477
BLSDKG_BuildQuorumVerificationVectors_parallel_400, 5, 3, 6.08894, 0.292929, 0.757811, 0.295639
BLSDKG_BuildQuorumVerificationVectors_simple_10, 5, 2500, 5.15814, 0.000399814, 0.000440104, 0.000405492
BLSDKG_BuildQuorumVerificationVectors_simple_100, 5, 25, 4.81259, 0.0381607, 0.0389298, 0.0384625
BLSDKG_BuildQuorumVerificationVectors_simple_400, 5, 1, 3.21568, 0.626318, 0.665364, 0.642478
BLSDKG_VerifyContributionShares_aggregated_10, 5, 70, 4.48856, 0.0126555, 0.0131477, 0.0127854
BLSDKG_VerifyContributionShares_aggregated_100, 5, 2, 4.77096, 0.435957, 0.531176, 0.471338
BLSDKG_VerifyContributionShares_aggregated_400, 5, 1, 20.0569, 3.47841, 4.28962, 4.07697
BLSDKG_VerifyContributionShares_parallel_10, 5, 200, 6.57893, 0.00581002, 0.00804017, 0.00599347
BLSDKG_VerifyContributionShares_parallel_100, 5, 2, 5.09107, 0.486311, 0.587155, 0.487484
BLSDKG_VerifyContributionShares_parallel_400, 5, 1, 41.6826, 8.06761, 8.63975, 8.37928
BLSDKG_VerifyContributionShares_parallel_aggregated_10, 5, 150, 5.37699, 0.0066828, 0.00818343, 0.00685109
BLSDKG_VerifyContributionShares_parallel_aggregated_100, 5, 4, 4.97799, 0.237098, 0.268741, 0.247238
BLSDKG_VerifyContributionShares_parallel_aggregated_400, 5, 1, 10.2419, 1.82211, 2.13127, 2.10769
BLSDKG_VerifyContributionShares_simple_10, 5, 70, 4.09386, 0.0109543, 0.0126837, 0.0111479
BLSDKG_VerifyContributionShares_simple_100, 5, 1, 4.90478, 0.958188, 1.02932, 0.97319
BLSDKG_VerifyContributionShares_simple_400, 5, 1, 77.9934, 15.3448, 15.8714, 15.5736
BLS_PubKeyAggregate_Normal, 5, 140000, 4.95579, 7.05378e-06, 7.15263e-06, 7.06455e-06
BLS_SecKeyAggregate_Normal, 5, 800000, 5.14803, 1.28203e-06, 1.28963e-06, 1.28829e-06
BLS_Sign_Normal, 5, 600, 4.40597, 0.00145638, 0.00149072, 0.00146491
BLS_SignatureAggregate_Normal, 5, 100000, 5.18232, 1.03104e-05, 1.04003e-05, 1.03666e-05
BLS_Verify_Batched, 5, 500, 4.91768, 0.00187499, 0.00212658, 0.00196962
BLS_Verify_BatchedParallel, 5, 1000, 4.66455, 0.000916545, 0.000943917, 0.000936157
BLS_Verify_LargeAggregatedBlock100, 5, 7, 4.93739, 0.140928, 0.141234, 0.14101
BLS_Verify_LargeAggregatedBlock1000, 5, 1, 7.02685, 1.39841, 1.41383, 1.40514
BLS_Verify_LargeAggregatedBlock1000PreVerified, 5, 5, 3.75674, 0.149865, 0.150549, 0.150381
BLS_Verify_LargeBlock100, 5, 3, 4.57215, 0.30382, 0.305469, 0.304973
BLS_Verify_LargeBlock1000, 5, 1, 15.358, 3.0358, 3.12992, 3.06139
BLS_Verify_LargeBlockSelfAggregated100, 5, 7, 5.02241, 0.142189, 0.146418, 0.142672
BLS_Verify_LargeBlockSelfAggregated1000, 5, 1, 7.0435, 1.39884, 1.42163, 1.40738
BLS_Verify_Normal, 5, 350, 5.36255, 0.00304772, 0.00310296, 0.00305239
```

This PR:
```
# Benchmark, evals, iterations, total, min, max, median
BLSDKG_BuildQuorumVerificationVectors_parallel_10, 5, 12000, 5.13755, 7.86945e-05, 0.000105131, 7.97337e-05
BLSDKG_BuildQuorumVerificationVectors_parallel_100, 5, 170, 4.77972, 0.0055425, 0.00579877, 0.00559518
BLSDKG_BuildQuorumVerificationVectors_parallel_400, 5, 8, 4.37557, 0.108588, 0.110115, 0.109562
BLSDKG_BuildQuorumVerificationVectors_simple_10, 5, 11000, 4.94715, 8.95659e-05, 9.05387e-05, 8.99815e-05
BLSDKG_BuildQuorumVerificationVectors_simple_100, 5, 110, 4.86016, 0.0087459, 0.00904099, 0.00879282
BLSDKG_BuildQuorumVerificationVectors_simple_400, 5, 6, 5.04596, 0.167772, 0.16903, 0.168014
BLSDKG_VerifyContributionShares_aggregated_10, 5, 70, 4.42195, 0.0124147, 0.012875, 0.0126996
BLSDKG_VerifyContributionShares_aggregated_100, 5, 2, 4.81077, 0.412286, 0.531059, 0.485066
BLSDKG_VerifyContributionShares_aggregated_400, 5, 1, 17.8261, 3.37136, 3.69313, 3.6722
BLSDKG_VerifyContributionShares_parallel_10, 5, 200, 5.72349, 0.00564236, 0.0058847, 0.00567855
BLSDKG_VerifyContributionShares_parallel_100, 5, 2, 4.97458, 0.491292, 0.505288, 0.49691
BLSDKG_VerifyContributionShares_parallel_400, 5, 1, 40.4173, 7.87268, 8.36335, 8.00442
BLSDKG_VerifyContributionShares_parallel_aggregated_10, 5, 150, 4.91287, 0.00636193, 0.00662992, 0.00659231
BLSDKG_VerifyContributionShares_parallel_aggregated_100, 5, 4, 4.73395, 0.221334, 0.265993, 0.232032
BLSDKG_VerifyContributionShares_parallel_aggregated_400, 5, 1, 9.4278, 1.78727, 1.94501, 1.90603
BLSDKG_VerifyContributionShares_simple_10, 5, 70, 3.83386, 0.0108497, 0.0111273, 0.0109168
BLSDKG_VerifyContributionShares_simple_100, 5, 1, 5.29861, 0.981199, 1.15943, 1.05667
BLSDKG_VerifyContributionShares_simple_400, 5, 1, 84.0435, 15.4488, 18.2057, 16.3801
BLS_PubKeyAggregate_Normal, 5, 700000, 4.86448, 1.38583e-06, 1.39766e-06, 1.38856e-06
BLS_SecKeyAggregate_Normal, 5, 1300000, 5.23721, 7.93552e-07, 8.2957e-07, 7.99736e-07
BLS_Sign_Normal, 5, 600, 4.3102, 0.00142861, 0.00145476, 0.00143372
BLS_SignatureAggregate_Normal, 5, 300000, 4.91937, 3.2663e-06, 3.29615e-06, 3.27328e-06
BLS_Verify_Batched, 5, 500, 4.95215, 0.00179908, 0.00225202, 0.00191845
BLS_Verify_BatchedParallel, 5, 1000, 4.68432, 0.000929385, 0.000955866, 0.000931772
BLS_Verify_LargeAggregatedBlock100, 5, 7, 5.04091, 0.141141, 0.14829, 0.14333
BLS_Verify_LargeAggregatedBlock1000, 5, 1, 7.05112, 1.39728, 1.41846, 1.4117
BLS_Verify_LargeAggregatedBlock1000PreVerified, 5, 7, 5.06663, 0.144226, 0.146193, 0.144423
BLS_Verify_LargeBlock100, 5, 3, 4.62824, 0.30777, 0.309262, 0.308754
BLS_Verify_LargeBlock1000, 5, 1, 15.4331, 3.0698, 3.10921, 3.08757
BLS_Verify_LargeBlockSelfAggregated100, 5, 7, 4.98629, 0.142051, 0.143493, 0.142174
BLS_Verify_LargeBlockSelfAggregated1000, 5, 1, 7.13337, 1.4119, 1.46093, 1.41687
BLS_Verify_Normal, 5, 350, 5.39525, 0.00306339, 0.00312148, 0.00307178
```

Based on #4012 atm.

### CODE REVIEW

**Review by xdustinface — CHANGES_REQUESTED** [review]
> IMO doing this in `GetHash()` only and only when it's actually needed should be enough.

Agree 👍 Thoughts about [suggestions-4013](https://github.com/xdustinface/dasha/tree/suggestions-4013)?

**Comment by UdjinM6** [comment]
> > IMO doing this in `GetHash()` only and only when it's actually needed should be enough.
> 
> Agree 👍 Thoughts about [suggestions-4013](https://github.com/xdustinface/dash/tree/suggestions-4013)?

Not sure I like having a separate bool to indicate the `Null`-ness of the cached hash instead of simply calling `IsNull()` on it tbh, feels like an easy way to mess up 🙂 Plus, there are no performance gains either (at least as far as I can see from my local benchmarks), so I'd rather avoid that. Dropping `UpdateHash()`s makes sense though, added.

**Comment by xdustinface** [comment]
>  simply calling SetNull/IsNull()

Hm ok agree, it's code-wise the cleaner solution and i also don't see any noticeable improvement in benchs, better to leave it as it is. But still keep in mind, thats a 32x byte set/compare.. now think a moment about how many extra bytes this are in 5 years 😄

**Review by xdustinface — APPROVED** [review]
utACK

File: src/bench/bls.cpp [hunk 1]
```diff
@@ -359,9 +359,9 @@ static void BLS_Verify_BatchedParallel(benchmark::State& state)
     }
 }
 
-BENCHMARK(BLS_PubKeyAggregate_Normal, 140 * 1000)
-BENCHMARK(BLS_SecKeyAggregate_Normal, 800 * 1000)
-BENCHMARK(BLS_SignatureAggregate_Normal, 100 * 1000)
+BENCHMARK(BLS_PubKeyAggregate_Normal, 700 * 1000)
+BENCHMARK(BLS_SecKeyAggregate_Normal, 1300 * 1000)
+BENCHMARK(BLS_SignatureAggregate_Normal, 300 * 1000)
```
Comment (PastaPastaPasta) [inline src/bench/bls.cpp:364]: Why change this here?

File: src/bench/bls.cpp [hunk 2]
```diff
@@ -370,6 +370,6 @@ BENCHMARK(BLS_Verify_LargeBlockSelfAggregated100, 7)
 BENCHMARK(BLS_Verify_LargeBlockSelfAggregated1000, 1)
 BENCHMARK(BLS_Verify_LargeAggregatedBlock100, 7)
 BENCHMARK(BLS_Verify_LargeAggregatedBlock1000, 1)
-BENCHMARK(BLS_Verify_LargeAggregatedBlock1000PreVerified, 5)
+BENCHMARK(BLS_Verify_LargeAggregatedBlock1000PreVerified, 7)
```
Comment (PastaPastaPasta) [inline src/bench/bls.cpp:373]: same

File: src/bench/bls_dkg.cpp [hunk 3]
```diff
@@ -147,12 +147,12 @@ void CleanupBLSDkgTests()
...
-BENCH_BuildQuorumVerificationVectors(simple, 400, false, 1)
-BENCH_BuildQuorumVerificationVectors(parallel, 10, true, 3000)
-BENCH_BuildQuorumVerificationVectors(parallel, 100, true, 50)
-BENCH_BuildQuorumVerificationVectors(parallel, 400, true, 3)
+BENCH_BuildQuorumVerificationVectors(simple, 10, false, 11000)
+BENCH_BuildQuorumVerificationVectors(simple, 100, false, 110)
+BENCH_BuildQuorumVerificationVectors(simple, 400, false, 6)
+BENCH_BuildQuorumVerificationVectors(parallel, 10, true, 12000)
+BENCH_BuildQuorumVerificationVectors(parallel, 100, true, 170)
+BENCH_BuildQuorumVerificationVectors(parallel, 400, true, 8)
```
Comment (PastaPastaPasta) [inline src/bench/bls_dkg.cpp:155]: same

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
I'd like to drop benchmark count changes and then post new pre-post benchmarks with equal run counts

**Comment by UdjinM6** [comment]
@PastaPastaPasta 

> Why change this here?

Because benchmarks should be kept close to 1 second mark each (per evaluation) if possible, see https://github.com/dashpay/dash/blob/develop/src/bench/bench.h#L136-L138

> I'd like to drop benchmark count changes and then post new pre-post benchmarks with equal run counts

This PR with d3407b2 reverted (i.e. with old counts):
```
# Benchmark, evals, iterations, total, min, max, median
BLSDKG_BuildQuorumVerificationVectors_parallel_10, 5, 3000, 1.16562, 7.53368e-05, 8.36367e-05, 7.56441e-05
BLSDKG_BuildQuorumVerificationVectors_parallel_100, 5, 50, 1.40202, 0.00549386, 0.00579628, 0.00557647
BLSDKG_BuildQuorumVerificationVectors_parallel_400, 5, 3, 1.64643, 0.106558, 0.118435, 0.108393
BLSDKG_BuildQuorumVerificationVectors_simple_10, 5, 2500, 1.12774, 8.94384e-05, 9.06852e-05, 9.04272e-05
BLSDKG_BuildQuorumVerificationVectors_simple_100, 5, 25, 1.0996, 0.00870218, 0.0089259, 0.00877597
BLSDKG_BuildQuorumVerificationVectors_simple_400, 5, 1, 0.830553, 0.164213, 0.168017, 0.166152
BLSDKG_VerifyContributionShares_aggregated_10, 5, 70, 4.40524, 0.0123473, 0.0129525, 0.0124959
BLSDKG_VerifyContributionShares_aggregated_100, 5, 2, 4.52029, 0.409926, 0.496239, 0.455014
BLSDKG_VerifyContributionShares_aggregated_400, 5, 1, 18.2816, 3.39591, 3.85397, 3.69386
BLSDKG_VerifyContributionShares_parallel_10, 5, 200, 5.70737, 0.00564582, 0.00575834, 0.00570326
BLSDKG_VerifyContributionShares_parallel_100, 5, 2, 5.10614, 0.48898, 0.575816, 0.496705
BLSDKG_VerifyContributionShares_parallel_400, 5, 1, 40.3521, 8.00261, 8.13095, 8.07218
BLSDKG_VerifyContributionShares_parallel_aggregated_10, 5, 150, 5.12541, 0.00665255, 0.00700472, 0.00681193
BLSDKG_VerifyContributionShares_parallel_aggregated_100, 5, 4, 5.11267, 0.241041, 0.28616, 0.249254
BLSDKG_VerifyContributionShares_parallel_aggregated_400, 5, 1, 9.45805, 1.76284, 1.9471, 1.91327
BLSDKG_VerifyContributionShares_simple_10, 5, 70, 3.9014, 0.0110082, 0.011281, 0.0111315
BLSDKG_VerifyContributionShares_simple_100, 5, 1, 5.45545, 1.03455, 1.14073, 1.07651
BLSDKG_VerifyContributionShares_simple_400, 5, 1, 78.7225, 15.6368, 15.8353, 15.7648
BLS_PubKeyAggregate_Normal, 5, 140000, 0.958813, 1.35601e-06, 1.40488e-06, 1.36394e-06
BLS_SecKeyAggregate_Normal, 5, 800000, 3.34697, 8.33283e-07, 8.43286e-07, 8.34975e-07
BLS_Sign_Normal, 5, 600, 4.32459, 0.00143482, 0.00145369, 0.00144088
BLS_SignatureAggregate_Normal, 5, 100000, 1.62545, 3.21642e-06, 3.34225e-06, 3.22498e-06
BLS_Verify_Batched, 5, 500, 5.0861, 0.00179302, 0.00215789, 0.00208866
BLS_Verify_BatchedParallel, 5, 1000, 4.63183, 0.000906306, 0.000943805, 0.00092583
BLS_Verify_LargeAggregatedBlock100, 5, 7, 5.09773, 0.145133, 0.146212, 0.145503
BLS_Verify_LargeAggregatedBlock1000, 5, 1, 7.21984, 1.42212, 1.49494, 1.43035
BLS_Verify_LargeAggregatedBlock1000PreVerified, 5, 5, 3.6531, 0.145337, 0.146893, 0.146015
BLS_Verify_LargeBlock100, 5, 3, 4.61907, 0.307085, 0.309952, 0.30768
BLS_Verify_LargeBlock1000, 5, 1, 15.4607, 3.07065, 3.10484, 3.10168
BLS_Verify_LargeBlockSelfAggregated100, 5, 7, 5.07569, 0.144547, 0.145262, 0.145163
BLS_Verify_LargeBlockSelfAggregated1000, 5, 1, 7.24931, 1.42817, 1.483, 1.4367
BLS_Verify_Normal, 5, 350, 5.67469, 0.00309775, 0.00351677, 0.00311346
```

**Review by PastaPastaPasta — APPROVED** [review]
Hmm, okay, I guess that's fair :+1: 

utACK

**Pull request merged**